### PR TITLE
FOGL-2309 - Poll mode Python Playback plugin

### PR DESCRIPTION
**Pull request by amarendra-dianomic**
Seems that South C server does not support single quote ('), double quote (") and backward slash (\\) in string data. Tested with below csv data:
```
TestInt,TestFloat,TestStr
2,2.1," !#$%&()*+,-./0123456789:;<=>?@"
0,0.0,"ABCDEFGHIJKLMNOPQRSTUVWXYZ[]^_`"
-2,-2.1,"abcdefghijklmnopqrstuvwxyz{|}~"
```

Also, C++ server is not able to support Burst mode in which nested readings are captured:
```
Burst size: 3
{
	'readings': {
		'data': [{
			'TestFloat': 2.1,
			'TestInt': 2,
			'TestStr': ' !#$%&()*+,-./0123456789:;<=>?@'
		}, {
			'TestFloat': 0.0,
			'TestInt': 0,
			'TestStr': 'ABCDEFGHIJKLMNOPQRSTUVWXYZ[]^_`'
		}, {
			'TestFloat': -2.1,
			'TestInt': -2,
			'TestStr': 'abcdefghijklmnopqrstuvwxyz{|}~'
		}]
	},
	'asset': 'ample2',
	'key': '513d31f7-bdd4-4bf6-8ca2-4990863a1d7a',
	'timestamp': '2019-01-17 14:43:26.339795+05:30'
}

Burst size: 1
{
	'readings': {
		'data': [{
			'TestFloat': 0.0,
			'TestInt': 0,
			'TestStr': 'ABCDEFGHIJKLMNOPQRSTUVWXYZ[]^_`'
		}]
	},
	'asset': 'ample2',
	'key': '825ffba1-7f0e-4445-ad5b-49fa1f6d06ea',
	'timestamp': '2019-01-17 14:43:51.339872+05:30'
}
{
	'readings': {
		'data': [{
			'TestFloat': 0.0,
			'TestInt': 0,
			'TestStr': 'ABCDEFGHIJKLMNOPQRSTUVWXYZ[]^_`'
		}]
	},
	'asset': 'ample2',
	'key': '825ffba1-7f0e-4445-ad5b-49fa1f6d06ea',
	'timestamp': '2019-01-17 14:43:51.339872+05:30'
}
{
	'readings': {
		'data': [{
			'TestFloat': -2.1,
			'TestInt': -2,
			'TestStr': 'abcdefghijklmnopqrstuvwxyz{|}~'
		}]
	},
	'asset': 'ample2',
	'key': 'f7216825-b4d4-4893-9591-a7fa6a6f6028',
	'timestamp': '2019-01-17 14:43:53.339849+05:30'
}

```

### CODE REVIEW

**Comment by amarendra-dianomic** [comment]
This plugin is working fine with Python South service but is not responding to South C service. Even plugin_init() is not responding to shim layer. Needs further investigation.

File: python/foglamp/plugins/south/playback/playback.py [hunk 1]
```diff
@@ -40,6 +38,11 @@
         'default': 'playback',
         'readonly': 'true'
     },
+    'pollInterval': {
```
Comment (praveen-garg) [inline python/foglamp/plugins/south/playback/playback.py]: remove this! 

Also check plugin_reconfigure and remove restart key.

File: python/foglamp/plugins/south/playback/playback.py [hunk 2]
```diff
@@ -252,11 +255,9 @@ def plugin_reconfigure(handle, new_config):
         'repeatLoop' in diff:
         plugin_shutdown(handle)
         new_handle = plugin_init(new_config)
-        new_handle['restart'] = 'yes'
         _LOGGER.info("Restarting playback plugin due to change in configuration key [{}]".format(', '.join(diff)))
```
Comment (praveen-garg) [inline python/foglamp/plugins/south/playback/playback.py]: do we really need to check the diff? and Restart will happen even in poll mode?

Comment (praveen-garg) [reply python/foglamp/plugins/south/playback/playback.py]: Though, we should log the diff on reconfigure

**Review by Singhal-Vaibhav — CHANGES_REQUESTED** [review]
Issues in change callback:
```
Jan 15 15:27:21 nerd047 FogLAMP[13008] INFO: playback: foglamp.plugins.south.playback.playback: playback plugin shut down.
Jan 15 15:27:21 nerd047 FogLAMP[13008] INFO: playback: foglamp.plugins.south.playback.playback: Restarting playback plugin due to change in configuration key [csvFilename]
Jan 15 15:27:21 nerd047 play[13008]: FATAL: Signal 6 (Aborted) trapped:
Jan 15 15:27:21 nerd047 play[13008]: FATAL: (0) 0             0x41374d handler(int) + 61---------
Jan 15 15:27:21 nerd047 play[13008]: FATAL: (1) 1       0x7f2f25fbd4b0 /lib/x86_64-linux-gnu/libc.so.6(+0x354b0) [0x7f2f25fbd4b0]---------
Jan 15 15:27:21 nerd047 play[13008]: FATAL: (2) 2       0x7f2f25fbd428 gsignal + 56---------
Jan 15 15:27:21 nerd047 play[13008]: FATAL: (3) 3       0x7f2f25fbf02a abort + 362---------
Jan 15 15:27:21 nerd047 play[13008]: FATAL: (4) 4       0x7f2f265f784d __gnu_cxx::__verbose_terminate_handler() + 365---------
Jan 15 15:27:21 nerd047 play[13008]: FATAL: (5) 5       0x7f2f265f56b6 /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x8d6b6) [0x7f2f265f56b6]---------
Jan 15 15:27:21 nerd047 play[13008]: FATAL: (6) 6       0x7f2f265f5701 /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x8d701) [0x7f2f265f5701]---------
Jan 15 15:27:21 nerd047 play[13008]: FATAL: (7) 7       0x7f2f26620d38 /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xb8d38) [0x7f2f26620d38]---------
Jan 15 15:27:21 nerd047 play[13008]: FATAL: (8) 8       0x7f2f270136ba /lib/x86_64-linux-gnu/libpthread.so.0(+0x76ba) [0x7f2f270136ba]---------
Jan 15 15:27:21 nerd047 play[13008]: FATAL: (9) 9       0x7f2f2608f41d clone + 109---------
Jan 15 15:27:21 nerd047 FogLAMP[11557] ERROR: change_callback: foglamp.services.core.interest_registry.change_callback: Unable to notify microservice with uuid b8ca1c32-465f-4898-a397-08e1e435104a due to exception: None#012Traceback (most recent call last):#012  File "/home/foglamp/Development/FogLAMP/python/foglamp/services/core/interest_registry/change_callback.py", line 59, in run#012    async with session.post(url, data=json.dumps(payload, sort_keys=True), headers=headers) as resp:#012  File "/home/foglamp/.local/lib/python3.5/site-packages/aiohttp/client.py", line 855, in __aenter__#012    self._resp = await self._coro#012  File "/home/foglamp/.local/lib/python3.5/site-packages/aiohttp/client.py", line 391, in _request#012    await resp.start(conn)#012  File "/home/foglamp/.local/lib/python3.5/site-packages/aiohttp/client_reqrep.py", line 757, in start#012    message, payload = await self._protocol.read()#012  File "/home/foglamp/.local/lib/python3.5/site-packages/aiohttp/streams.py", line 543, in read#012    await self._waiter#012  File "/usr/lib/python3.5/asyncio/futures.py", line 380, in __iter__#012    yield self  # This tells Task to wait for completion.#012  File "/usr/lib/python3.5/asyncio/tasks.py", line 309, in _wakeup#012    future.result()#012  File "/usr/lib/python3.5/asyncio/futures.py", line 293, in result#012    raise self._exception#012aiohttp.client_exceptions.ServerDisconnectedError: None
```

File: python/foglamp/plugins/south/playback/playback.py [hunk 3]
```diff
@@ -159,7 +157,7 @@ def plugin_info():
     return {
         'name': 'Playback',
         'version': '1.0',
```
Comment (ashish-jabble) [inline python/foglamp/plugins/south/playback/playback.py]: 2.0

**Review by MarkRiddoch — APPROVED** [review]

File: python/foglamp/plugins/south/playback/playback.py [hunk 4]
```diff
@@ -158,8 +156,8 @@ def plugin_info():
     """
     return {
         'name': 'Playback',
-        'version': '1.0',
-        'mode': 'async',
+        'version': '2.0',
```
Comment (praveen-garg) [inline python/foglamp/plugins/south/playback/playback.py]: how about  `VERSION.south.` file that we are using for make_deb?! for Sinusoid we used `2.0.0` so let's be consistent.

Also we should backup current develop (async version) in `async_backup` branch and protect that.

**Comment by amarendra-dianomic** [comment]
@Singhal-Vaibhav a bug with South C server caused the issue reported by you above. It has been fixed vide FOGL-2316. Pl check.

**Comment by amarendra-dianomic** [comment]
Checked with below CSV (FOGL-1000):
```
TestInt,TestFloat,TestStr,ts
2,2.1," !#$%&()*+,-./0123456789:;<=>?@",19-Jan-16 15:25:35.123
0,0.0,"ABCDEFGHIJKLMNOPQRSTUVWXYZ[]^_`",19-Jan-16 15:25:36.123
-2,-2.1,"abcdefghijklmnopqrstuvwxyz{|}~",19-Jan-16 15:26:41.143
-2,-2e11,"abcdefghijklmnopqrstuvwxyz{|}~",19-Jan-16 15:26:42.143
-2,1e-2,"abcdefghijklmnopqrstuvwxyz{|}~",19-Jan-16 15:26:45.143
-9223372036854775808,1E+37,"abcdefghijklmnopqrstuvwxyz{|}~",19-Jan-16 15:26:47.143
9223372036854775807,1E-37,"abcdefghijklmnopqrstuvwxyz{|}~",19-Jan-16 15:26:49.143
10000000000000000000000000000000000000,0.0000000000000000000000000000000000001,xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx,19-Jan-16 15:26:51.143
```

and below is the data ingested into readings table:
```
[
    {
        "asset_code": "sample",
        "id": 43,
        "read_key": "d39fdbc8-9777-4dbf-8c2e-aee20cd31933",
        "reading": "{\"TestStr\":\" !#$%&()*+,-./0123456789:;<=>?@\",\"TestFloat\":2.1,\"TestInt\":2}",
        "ts": "2019-01-29 08:05:07.961",
        "user_ts": "2019-01-29 08:05:04.969068+00:00"
    },
    {
        "asset_code": "sample",
        "id": 44,
        "read_key": "5211fb83-dd91-416f-b0b2-74e7f036196d",
        "reading": "{\"TestStr\":\"ABCDEFGHIJKLMNOPQRSTUVWXYZ[]^_`\",\"TestFloat\":0,\"TestInt\":0}",
        "ts": "2019-01-29 08:05:13.169",
        "user_ts": "2019-01-29 08:05:07.966965+00:00"
    },
    {
        "asset_code": "sample",
        "id": 45,
        "read_key": "820ec117-0b10-43ff-b2f5-fbf516395f9e",
        "reading": "{\"TestStr\":\"abcdefghijklmnopqrstuvwxyz{|}~\",\"TestFloat\":-2.1,\"TestInt\":-2}",
        "ts": "2019-01-29 08:06:18.255",
        "user_ts": "2019-01-29 08:06:13.966646+00:00"
    },
    {
        "asset_code": "sample",
        "id": 46,
        "read_key": "4320aaec-8d24-416a-8ff7-5e3be4190923",
        "reading": "{\"TestStr\":\"abcdefghijklmnopqrstuvwxyz{|}~\",\"TestFloat\":-200000000000,\"TestInt\":-2}",
        "ts": "2019-01-29 08:06:18.255",
        "user_ts": "2019-01-29 08:06:16.966651+00:00"
    },
    {
        "asset_code": "sample",
        "id": 47,
        "read_key": "e320e171-2ef9-48bf-ac1e-0d7ed8411a98",
        "reading": "{\"TestStr\":\"abcdefghijklmnopqrstuvwxyz{|}~\",\"TestFloat\":0.01,\"TestInt\":-2}",
        "ts": "2019-01-29 08:06:23.326",
        "user_ts": "2019-01-29 08:06:20.966579+00:00"
    },
    {
        "asset_code": "sample",
        "id": 48,
        "read_key": "8f5cc5e2-07ea-4077-b895-260e589e464e",
        "reading": "{\"TestStr\":\"abcdefghijklmnopqrstuvwxyz{|}~\",\"TestFloat\":1e37,\"TestInt\":-9223372036854775808}",
        "ts": "2019-01-29 08:06:28.402",
        "user_ts": "2019-01-29 08:06:23.967031+00:00"
    },
    {
        "asset_code": "sample",
        "id": 49,
        "read_key": "f41f15f2-2dfb-45d9-a0da-440e55ae40cf",
        "reading": "{\"TestStr\":\"abcdefghijklmnopqrstuvwxyz{|}~\",\"TestFloat\":1e-37,\"TestInt\":9223372036854775807}",
        "ts": "2019-01-29 08:06:33.503",
        "user_ts": "2019-01-29 08:06:29.966846+00:00"
    },
    {
        "asset_code": "sample",
        "id": 50,
        "read_key": "e787ecfa-3148-40d6-a43a-ad74dc3bd1e6",
        "reading": "{\"TestStr\":\"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\",\"TestFloat\":1e-37,\"TestInt\":68739955140067328}",
        "ts": "2019-01-29 08:06:38.579",
        "user_ts": "2019-01-29 08:06:36.967133+00:00"
    }
]
```
Therefore,
- Characters single quote('), double quote(") and backslash(\) are not supported
- Integers such as 10000000000000000000000000000000000000 are not supported
- Floating point numbers will get converted to decimal formats, wherever possible otherwise floating point notation will be used. Float 0.0 will be ingested as 0.

**Comment by praveen-garg** [comment]
We should:
1.  remove 'data' key  in burst mode:

 ```
'readings': [{
			'TestFloat': -2.1,
			'TestInt': -2,
			'TestStr': 'abcdefghijklmnopqrstuvwxyz{|}~'
		    },
                    {
			'TestFloat': -2.2,
			'TestInt': -3,
			'TestStr': 'abcdefghijklmnopqrstuvwxyz{|}~'
		    }]
```
2. handle float correctly (not sure if 0.0 converts to 0 at the plugin or at ingest level?! and what happens with value like 2.0?)
  
```
ast.literal_eval("0.0")
0.0
type(ast.literal_eval("0.0"))
<class 'float'>
```

**Comment by praveen-garg** [comment]
> remove 'data' key in burst mode:

@amarendra-dianomic  for now let's bring back as C ingest has some type checking / mapping overheads.

**Review by praveen-garg — APPROVED** [review]

**Comment by praveen-garg** [comment]
Closing in favor of #5

**Pull request closed without merging**